### DISCUSSION TITLE
Proxy server with built in bot not working

### DISCUSSION THREAD

**Discussion by HumanDuck23** [PrismarineJS/mineflayer#2434] [Q&A] [answered]
I'm trying to make a proxy server with a bot built in. The user connects to the proxy running on their localhost, which then connects them to a server of their choice. This part works fine, I'm using the example script https://github.com/PrismarineJS/node-minecraft-protocol/tree/master/examples/proxy. What I want to add to this is a mineflayer bot. I want it to be able to (for example) move as if the player was moving on their normal client. The current issue I'm facing is that the user can't move properly (constant lagbacks, if I try and fly I slowly get teleported all the way down every few seconds). When I tell the bot to move It jumps forward every few seconds instead of having smooth movement.

The only code I've added myself is this (after `targetClient` is created in the example script)
```js
bot = mineflayer.createBot({
    client: targetClient
})
```
This already prevents the user from playing normally, but I also can't control the bot properly. Is it possible to do what I'm trying to do here?

**Comment by rom1504** [comment, marked as answer]
That would be cool to do. That has been discussed many times.
You will have to do many more changes
There is 2 sides you need to handle:
* Receiving packets by the Minecraft server: you need to make sure only one
position loop is sent by the proxy client (Minecraft client sends position
to the server every tick, either mineflayer either the vanilla client
should send them, not both)
* Receiving packets by the Minecraft client: you need to make sure all
information done by the bot is transmitted to the vanilla client somehow.
That will require extra hacking as in normal gameplay the server assumes
the client already knows where it is. In your case the client doesn't know,
so you would for example have to send teleport packet every time the
mineflayer bot moves


Movement is only one of the things that would need taking care of. Block
updates is another, and inventory is another.

This is a big project but I think it could be pretty cool.

Another way you can do at this is using prismarine-web-client which already
has mineflayer plugged in and which is an UI that allows controlling the
bot.

**Reply by HumanDuck23** [reply]
> That would be cool to do. That has been discussed many times. You will have to do many more changes There is 2 sides you need to handle: * Receiving packets by the Minecraft server: you need to make sure only one position loop is sent by the proxy client (Minecraft client sends position to the server every tick, either mineflayer either the vanilla client should send them, not both) * Receiving packets by the Minecraft client: you need to make sure all information done by the bot is transmitted to the vanilla client somehow. That will require extra hacking as in normal gameplay the server assumes the client already knows where it is. In your case the client doesn't know, so you would for example have to send teleport packet every time the mineflayer bot moves Movement is only one of the things that would need taking care of. Block updates is another, and inventory is another. This is a big project but I think it could be pretty cool. Another way you can do at this is using prismarine-web-client which already has mineflayer plugged in and which is an UI that allows controlling the bot.
> […](#)
> On Fri, Jan 28, 2022, 06:51 HumanDuck23 ***@***.***> wrote: I'm trying to make a proxy server with a bot built in. The user connects to the proxy running on their localhost, which then connects them to a server of their choice. This part works fine, I'm using the example script https://github.com/PrismarineJS/node-minecraft-protocol/tree/master/examples/proxy. What I want to add to this is a mineflayer bot. I want it to be able to (for example) move as if the player was moving on their normal client. The current issue I'm facing is that the user can't move properly (constant lagbacks, if I try and fly I slowly get teleported all the way down every few seconds). When I tell the bot to move It jumps forward every few seconds instead of having smooth movement. The only code I've added myself is this (after targetClient is created in the example script) bot = mineflayer.createBot({ client: targetClient}) This already prevents the user from playing normally, but I also can't control the bot properly. Is it possible to do what I'm trying to do here? — Reply to this email directly, view it on GitHub <#2434>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAR437T2PAIKNOSA77NDFODUYIVHBANCNFSM5M73SGWQ> . You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

Thanks for the info! I'll try the web client once I get home, and maybe I could try doing it the other way when I have time for a big project.

**Reply by HumanDuck23** [reply]
> 

I've thought about this for a while now, and come up with the following to describe how this should work:

![image](https://user-images.githubusercontent.com/44237686/151617495-15cc2a4b-fc6c-4bac-80af-8a10b6b8bf1b.png)

Now the part that really confuses me is how the proxy server is supposed to know which packets coming from the bot **and** the player it should send on. If I only send the ones from the bot, only the bot will work, the player won't be able to interact with the game (and the other way around). I have no idea how I'm supposed to be able to make both the player and the bot be able to interact with the game. Any ideas?

Thanks

**Reply by rom1504** [reply]
Yes I think you need the bot should have the upper hand and take as input the action from the vanilla client

**Reply by HumanDuck23** [reply]
> Yes I think you need the bot should have the upper hand and take as input the action from the vanilla client

So maybe something like this:

Vanilla client packets go to the bot, and the bot can either override them with its own packets or send them on to the proxy?

**Reply by HumanDuck23** [reply]
> Yes I think you need the bot should have the upper hand and take as input the action from the vanilla client

One more question: I was able to make the bot move instead of the client and send it regular position updates, but now the client doesn't play the walking animation. Is there a way to make it play the animation with some kind of packet?

**Reply by IceTank** [reply]
I already implemented something similar to this. https://github.com/IceTank/mineflayer-proxy-inspector What this package does it do the following: 
1. Create a bot (client A) that joins a server (server A)
2. Create a server (server B) any client can connect to (client B)

If a client connects to the server being hosted it recreates the world state, entities and game state by the information the bot has stored using [mcproxy](https://github.com/rob9315/mcproxy). Then it creates a link where all packets send from the server A are also send the client B. But packet send from client B to server B are blocked from going to server A. This keeps client B up to date with the state of server A and client A while not sending packets to server A when client A is already sending packets to server A.
The package has an in build command `$link` that once typed in chat will switch the packet routing around. All packets send by client B are send to server A. All packets send by client A are blocked from being send to server A, but it still receives all packets send by server A to keep the underlying mineflayer bot in synch. 
The command `$unlick` does the same in reverse. Only that now client A is in control again.

Some limitations are: It only works for version 1.12.2. As mcproxy only supports the re-creations of packets for the game world from versions 1.12.2. This could be fixed in the future by adding packets for other versions. As the hosted server B is hosted in offline mode all skins are also dependent on the given player uuid. This might be fixable in the future. This package also uses a fork of mcproxy to add better packet handling when intercepting or changing packets in transit. But due to a workaround for a crashing issue this fork is unable to be merged into mcproxy at the moment.

There are also some other features: like spawning a fake player representing the underlying mineflayer bot. This is so you can see the underlying mineflayer bot walking around and interacting with stuff. Once you take control this fake player is automatically de-spawned. Another feature that is not in the repo right now is a special packet send to the client B that tells it to inspect the spawned fake player. This causes your game client to have its camera looked into the same view as the fake player. So you are seeing the world through the eyes of the bot even when he moves around and interacting with the world.

**Reply by HumanDuck23** [reply]
That's really cool! Does it **only** work for version 1.12.2 or also for lower versions? My bot is intended to run on 1.8.9 but it probably doesn't matter a lot.

**Reply by IceTank** [reply]
You have to check. The limiting factor are the hardcoded packets in mcproxy and maybe in my package. Depending on how much the packets have changed it should be not that difficult to port it to other versions.
